components/client/src/main/java/com/hotels/styx/client/netty/connectionpool/NettyConnection.java

```diff
@@ -54,13 +59,28 @@
      * @param requestOperationFactory used to create operation objects that send http requests via this connection
      */
     @VisibleForTesting
-    public NettyConnection(Origin origin, Channel channel, HttpRequestOperationFactory requestOperationFactory) {
+    public <T> NettyConnection(Origin origin, Channel channel, HttpRequestOperationFactory requestOperationFactory, HttpConfig httpConfig, SslContext sslContext) {
         this.origin = requireNonNull(origin);
         this.channel = requireNonNull(channel);
         this.requestOperationFactory = requestOperationFactory;
         this.channel.pipeline().addLast(new TimeToFirstByteHandler(this));
         this.channel.closeFuture().addListener(future ->
                 listeners.announce().connectionClosed(NettyConnection.this));
+
+        addChannelHandlers(channel, httpConfig, sslContext);
+    }
+
+    private static void addChannelHandlers(Channel channel, HttpConfig httpConfig, SslContext sslContext) {
```

Comment:
If this is purely for testing, perhaps we should let the test specify which `ChannelHandler`s to add, so that we avoid adding too much code that is only used during tests?

Reply:
This change is necessary to allow API consumer to choose whether to send a request over `http` or secure `https` protocol.

The problem we had is that `NettyConnectionFactory` can only provide channels that are either bound to `http` or `https` protocol. Therefore a factory that creates `http` channels cannot be used to send `https` traffic. This is fine for the styx proxy but doesn't work well with a stand-alone client.

This changes so that `NettyConnectionFactory` consumer can choose if it wants a `http` or `https` channel. That is, the factory has just enough information to build "bare" netty channels, and the actual channel configuration is injected from the consumer in a `createConnection` call.

Reply:
Updated PR description with a rationale to this change.